python/ray/_private/metrics_agent.py

```diff
@@ -60,43 +66,109 @@ def name(self):
 Record = namedtuple("Record", ["gauge", "value", "tags"])
 
 
+class WorkerProxyExportState:
+    def __init__(self):
+        self._last_reported_time = time.time()
+        # view_name
+        # -> set of tag_vals tuple that this worker genereated.
+        # NOTE: We assume a tuple of tag_vals correspond to a
+        # single worker because it always contain WorkerId.
+        self._view_to_owned_tag_vals = defaultdict(set)
+
+    @property
+    def last_reported_time(self):
+        return self._last_reported_time
+
+    @property
+    def view_to_owned_tag_vals(self):
+        return self._view_to_owned_tag_vals
+
+    def update_last_reported_time(self):
+        self._last_reported_time = time.time()
+
+    def put_tag_vals(self, view_name: str, tag_vals: Tuple[str]):
+        self._view_to_owned_tag_vals[view_name].add(tag_vals)
+
+
 class MetricsAgent:
-    def __init__(self, metrics_export_address, metrics_export_port):
-        assert metrics_export_port is not None
-        # OpenCensus classes.
-        self.view_manager = stats_module.stats.view_manager
-        self.stats_recorder = stats_module.stats.stats_recorder
-        # Port where we will expose metrics.
-        self.metrics_export_port = metrics_export_port
+    def __init__(
+        self,
+        view_manager: ViewManager,
+        stats_recorder: StatsRecorder,
+        stats_exporter: StatsExporter = None,
+    ):
+        """A class to record and export metrics.
+
+        The class exports metrics in 2 different ways.
+        - Directly record and export metrics using OpenCensus.
+        - Proxy metrics from other core components
+            (e.g., raylet, GCS, core workers).
+
+        This class is thread-safe.
+        """
         # Lock required because gRPC server uses
         # multiple threads to process requests.
         self._lock = threading.Lock()
```

Comment:
do you know where is the grpc handler defined for this?

Reply:
this is not a handler. The handler is defined in `reporter_agent.py`